### PR TITLE
Avoid in-place updates for compute/infra nodes

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/openshift/openshift-azure/issues/281

@openshift/sig-azure

### CODE REVIEW

**Comment by kargakis** [comment]
Almost an hour for 3 masters (25m), 1 infra (6m), and 3 compute nodes (20m). Most time seems to be spent on deallocating masters and deleting compute/infra VMs.
<details>

```
INFO[2018-08-30T15:35:06+02:00] convert to internal                           resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] enrich                                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] read old config                               resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] merge old and new config                      resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] merging internal data models                  resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] validating internal data models               resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] generating configs                            resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] persist config                                resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] generating arm templates                      resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] write helpers                                 resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:06+02:00] creating/updating deployment                  resourceGroup=mkargakiosa
INFO[2018-08-30T15:35:18+02:00] waiting for deployment                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:38:05+02:00] initializing cluster                          resourceGroup=mkargakiosa
INFO[2018-08-30T15:38:08+02:00] draining master-000000                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:38:09+02:00] deallocating master-000000 (0)                resourceGroup=mkargakiosa
INFO[2018-08-30T15:41:12+02:00] updating master-000000                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:41:13+02:00] reimaging master-000000                       resourceGroup=mkargakiosa
INFO[2018-08-30T15:42:14+02:00] starting master-000000                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:44:15+02:00] waiting for master-000000 to be ready         resourceGroup=mkargakiosa
INFO[2018-08-30T15:46:12+02:00] draining master-000001                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:46:12+02:00] deallocating master-000001 (1)                resourceGroup=mkargakiosa
INFO[2018-08-30T15:49:15+02:00] updating master-000001                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:50:15+02:00] reimaging master-000001                       resourceGroup=mkargakiosa
INFO[2018-08-30T15:51:16+02:00] starting master-000001                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:53:17+02:00] waiting for master-000001 to be ready         resourceGroup=mkargakiosa
INFO[2018-08-30T15:55:22+02:00] draining master-000002                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:55:22+02:00] deallocating master-000002 (2)                resourceGroup=mkargakiosa
INFO[2018-08-30T15:58:27+02:00] updating master-000002                        resourceGroup=mkargakiosa
INFO[2018-08-30T15:58:27+02:00] reimaging master-000002                       resourceGroup=mkargakiosa
INFO[2018-08-30T15:59:28+02:00] starting master-000002                        resourceGroup=mkargakiosa
INFO[2018-08-30T16:01:29+02:00] waiting for master-000002 to be ready         resourceGroup=mkargakiosa
INFO[2018-08-30T16:03:20+02:00] setting ss-infra capacity to 2                resourceGroup=mkargakiosa
INFO[2018-08-30T16:06:23+02:00] waiting for infra-000001 to be ready          resourceGroup=mkargakiosa
INFO[2018-08-30T16:07:18+02:00] draining infra-000000                         resourceGroup=mkargakiosa
INFO[2018-08-30T16:07:25+02:00] deleting infra-000000                         resourceGroup=mkargakiosa
INFO[2018-08-30T16:09:27+02:00] setting ss-compute capacity to 4              resourceGroup=mkargakiosa
INFO[2018-08-30T16:12:31+02:00] waiting for compute-000006 to be ready        resourceGroup=mkargakiosa
INFO[2018-08-30T16:13:05+02:00] draining compute-000000                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:13:07+02:00] deleting compute-000000                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:16:08+02:00] setting ss-compute capacity to 4              resourceGroup=mkargakiosa
INFO[2018-08-30T16:18:10+02:00] waiting for compute-000007 to be ready        resourceGroup=mkargakiosa
INFO[2018-08-30T16:19:21+02:00] draining compute-000004                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:19:23+02:00] deleting compute-000004                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:22:25+02:00] setting ss-compute capacity to 4              resourceGroup=mkargakiosa
INFO[2018-08-30T16:25:28+02:00] waiting for compute-000008 to be ready        resourceGroup=mkargakiosa
INFO[2018-08-30T16:26:05+02:00] draining compute-000005                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:26:07+02:00] deleting compute-000005                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:09+02:00] starting health check                         resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:10+02:00] checking infrastructure service health        resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:10+02:00] checking  "prometheus-node-exporter"          resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:10+02:00] checking  "sync"                              resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:10+02:00] checking  "ovs"                               resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:10+02:00] checking  "sdn"                               resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "prometheus"                        resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "docker-registry"                   resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "registry-console"                  resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "router"                            resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "apiserver"                         resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "controller-manager"                resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:11+02:00] checking  "asb"                               resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:12+02:00] checking  "bootstrap-autoapprover"            resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:12+02:00] checking  "apiserver"                         resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:12+02:00] checking  "webconsole"                        resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:12+02:00] done checking infrastructure service health   resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:12+02:00] checking console health                       resourceGroup=mkargakiosa
INFO[2018-08-30T16:29:12+02:00] OK                                            resourceGroup=mkargakiosa
```
</details>

File: cmd/createorupdate/update.go [hunk 1]
```diff
@@ -80,70 +80,68 @@ func updatePlusOne(ctx context.Context, cs *api.OpenShiftManagedCluster, p api.P
...
 	}
 
+	// TODO: Filter out VMs that do not need to get upgraded. Should speed
+	// up retrying failed upgrades.
 	vmsBefore := map[string]struct{}{}
-	for _, vm := range vms {
+	for _, vm := range oldVMs {
 		vmsBefore[*vm.InstanceID] = struct{}{}
 	}
+	vmsUpgraded := map[string]struct{}{}
```
Comment (jim-minter) [inline cmd/createorupdate/update.go]: why is this needed, out of interest? any reason you can't add new vms into vmsBefore as you iterate?

**Comment by kargakis** [comment]
This avoids waiting for new VMs more than once

File: cmd/createorupdate/update.go (hunk 1 above)
Comment (kargakis) [reply cmd/createorupdate/update.go]: I think you are right. I used to iterate over vmsBefore but not anymore so I can reuse it.

**Comment by kargakis** [comment]
One more upgrade [here](http://pastebin.test.redhat.com/640504). Deallocating/deleting a VM is by far the longest action during the upgrade (3m per VM).
